### PR TITLE
Docs: quickstart v2 

### PR DESCRIPTION
**Pull request by kay-kim**
Ready for review.  Once the content is approved, will need to make changes to Console text so that changes can be coordinated.

https://preview.materialize.com/materialize/28864/get-started/quickstart/

### CODE REVIEW

**Comment by morsapaes** [comment]
Started reviewing the PR; but it's quite long, so I'll pick it back up next week — sorry! From a quick skim: @frankmcsherry and I spent some time thinking about how to refactor the previous version of the quickstart ([#21656](https://github.com/MaterializeInc/materialize/pull/21656)) to minimize the time to the aha! moment, because there was a high drop rate. I'd be weary of going back to that same approach of making the quickstart very verbose, and about teaching core concepts by explaining, rather than showing. Reading through the [epic](https://www.notion.so/materialize/Improve-Quickstart-e4f5e4143cca41269c67e59778df227f), I understand you have the opposite take, though!

Looking at the (limited) data: our quickstart completion rate seems...healthy. We also have hard evidence that [users like the refreshed version](https://materializeinc.slack.com/archives/CU7ELJ6E9/p1710419260391009). The rate of trials started vs. quickstarts started is what seems less good, and this will **not** be solved by changing the content.

**Comment by morsapaes** [comment]
To be clearer: I think that we're missing an accessible walk through the Materialize core concepts more than we're missing a more thorough quickstart. Assuming not everyone is going to open the quickstart, how can we ensure they get familiar with the core concepts? For users that jump directly to connecting their data, how do we prompt them to know what to do next? You already started working on this by improving the core concepts in the documentation, but we're missing the part where we help users connect the dots when they're completely skipping the quickstart (both in the documentation, and in the console).

File: doc/user/content/get-started/quickstart.md [hunk 1]
```diff
@@ -12,297 +12,764 @@ aliases:
...
+      ```mzsql
+      SUBSCRIBE TO (
+        SELECT *
+        FROM flippers
+      ) WITH (snapshot = false)
+      ;
+      ```
+
+1. In another [Materialize console](https://console.materialize.com/), insert
+   new data into the `known_flippers` table .
```
Comment (frankmcsherry) [inline doc/user/content/get-started/quickstart.md:537]: Oof. We had special tech for doing table insertions because the context shift was annoying. Should we repurpose that here, or use multiple browser windows?

**Review by frankmcsherry — COMMENTED** [review]
Bunch of comments left behind. I do think the quickstart is made less "quick" by these changes, and potentially less compelling. At least, the moments that I think of as "aha" moments are not built up to in the same way, and more time is spent on the mechanics of getting there. This is maybe totally fine, and perhaps even a dramatic improvement, but I think it is moving away from answering "I have 5 minutes; convince me to stick around longer" to something that feels more like an introduction to the details of our SQL commands and such. If this is intentional then great, and if it's unintentional perhaps we should discuss!

File: doc/user/content/get-started/quickstart.md (hunk 1 above)
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md:537]: Oh ... I wasn't aware of the special tech for table insertion.  We can definitely can get rid of the multiple windows if the console allows it.

File: doc/user/content/get-started/quickstart.md [hunk 2]
```diff
@@ -12,297 +12,764 @@ aliases:
...
+    | auction_house_progress | progress       | null | null       |
+    | auctions               | subsource      | 25cc | quickstart |
+    | bids                   | subsource      | 25cc | quickstart |
+    | organizations          | subsource      | 25cc | quickstart |
+    | users                  | subsource      | 25cc | quickstart |
     ```
 
-    For now, you'll focus on the `auctions` and `bids` data sets. Data will
-    be **continually produced** as you walk through the quickstart.
+    A [`subsource`](/sql/show-subsources) is a relation(table).
```
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md:136]: I just copied it since that's what our show subsources page stated.  But happy to just use table instead. (hence my parenthetical table).

**Comment by kay-kim** [comment]
As I address the feedback, I figured I'd also provide some context for the changes (as always, any and all feedback is welcome :heart: as I use these as data points -- some that may come into play now; some others as we get data that bears out the need for changes; some others as the product and the docs evolve. And there may be some that may never come into play as the product and the docs evolve). Iteration is the key to documentation; both in terms of an individual PR as well as the evolution of pages in the future (so, I strongly mean it that the feedback act as data points for me. Unfortunate for reviewers since as pages are refreshed ...  😉 ). For this PR, 

- It is intentional that it is a bit longer. For me, one aspect of the "quick" component in the quickstart is that I can start playing in Materialize right away. 
  - I agree with the sentiment in the feedback link provided by Marta, that having the Mat console where people can play around is great and want to build upon that.  Specifically, I think providing more examples to query in the console can provide our users with the opportunity to **start thinking about their queries** and what is possible in Materialize for them. So more their use-case aha moment (or even a "perhaps" moment), that is to spark an "Oh wait ... maybe this can work with our queries" or even "I bet this won't work with our query" (and have them try because we've given them some familiarity).
- I want to address the "Wait ... why am I creating a view and index instead of a materialized view? Isn't this company called Materialize? I could have sworn I saw some initial video/page on the site talking about their materialized view".  (Per Frank's feedback, I'll reword to try to make it clearer).
  - Per Frank's feedback, I'll reword to try to make it clearer.
  - And, agree with Marta that updates to our concepts pages are absolutely needed and will form a foundation. 
  - But, that will take a bit more time while currently, 
     - We have more people hitting our **Materialized Views** guide than are hitting either the **Quickstart** or the **index/views** concept pages.  
     - And more people hit **Quickstart** than index/views concept pages. So, while I'm hoping to revisit the quickstart in 3-4 months after we have made good inroads into our concept pages as well as reference pages, I think, if resources are stretched thin, the changes could buy us up to 9 months (fingers crossed).  But, as always, I'll be monitoring the metrics in order to act accordingly.
 
And yes ... I do think we will revisit this page as our docs evolve.  But, I take that as an opportunity to catch things we didn't before (or thought it was insignificant before, but having another look at it with benefit of time). 

All this to say in a nutshell: All opinions are welcome.  I may or may not act on them at this time.  Don't ever feel bad or feel the need to justify your opinions.  You can always leave arguments that strengthen your case, but definitely don't need to justify yourself to me.  And even if it's a feedback I took, and then, you change your mind after you see it, don't hesitate (because iteration).

File: doc/user/content/get-started/quickstart.md [hunk 3]
```diff
@@ -12,297 +12,764 @@ aliases:
...
+- Create views to verify that Materialize serves <red>consistent results</red>.
+
+{{< note >}}
 
-* **Interactivity**: get immediate responses from indexed warehouse relations and derived results.
+For the purpose of this quickstart, the results do not need to be persisted in
+durable storage. As such, the quickstart uses an index on a view to maintain
+up-to-date view results.
 
-* **Freshness**: watch results change immediately in response to *your* input changes.
```
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md:22]: I'll see if I can call these out more in the intro paragraph.  The reason for making these less in-your-face and incorporated into the "up-to-date" stuff in the paragraph is 2 fold:
- Users tend to be more interested in the actions that they themselves will do than the more passive action.
- Users' journey to this page will have included at least 1 page that will have called this out (be it our company site or docs landing page).  So, I'm not too worried about their prominence or lack thereof.

File: doc/user/content/get-started/quickstart.md [hunk 4]
```diff
@@ -12,297 +12,764 @@ aliases:
...
 
     ```mzsql
     CREATE SOURCE auction_house
     FROM LOAD GENERATOR AUCTION
     (TICK INTERVAL '1s', AS OF 100000)
     FOR ALL TABLES;
     ```
-1. Use the [`SHOW SOURCES`](/sql/show-sources/) command to get an idea of the data being generated:
+
+    `CREATE SOURCE` creates **multiple** sources when ingesting data from
```
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md:111]: Since they'll see subsource when they do Show Sources we won't be able to not introduce ... but, I took a second pass (will be in next patch).

File: doc/user/content/get-started/quickstart.md [hunk 5]
```diff
@@ -12,297 +12,764 @@ aliases:
...
+
+      As new data enters the system, above `SELECT` queries could be rerun to
+      get the recent results.
+
+      To efficiently serve queries that have no supporting indexes, Materialize
+      uses the same underlying mechanism it does for indexes. However, without
+      an index, the underlying work is discarded after the results are served.
+      By creating a view to save repeated queries and then <redb>indexing</redb>
+      the view, Materialize can <redb>maintain and incrementally update</redb>
+      view results in memory.
```
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md:209]: Will remove the first 2 sentences in the next patch.

File: doc/user/content/get-started/quickstart.md [hunk 6]
```diff
@@ -12,297 +12,764 @@ aliases:
...
+   ORDER BY bid_time DESC
+   LIMIT 10;
+   ```
+
+   Each time you query the view, you are running the underlying statement. Even
+   without an index, Materialize uses the same mechanics used by indexes to
+   optimize computations. However, without an index, the underlying work is
+   discarded after each query run.
+
+   In Materialize, you can create [**indexes**](/concepts/indexes/) on views to
```
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md:271]: Connecting.  Just a foreshadowing with the benefit that  if someone stops at this step (particularly in the console), at least there is a chance they see it.  And also, repetition can sometimes help make people pause (even if it is just to note that it's repeated) and increase likelihood of planting a seed.

File: doc/user/content/get-started/quickstart.md [hunk 7]
```diff
@@ -12,297 +12,764 @@ aliases:
...
-    Like in other SQL databases, a view in Materialize is just an alias for the
-    embedded `SELECT` statement; results are computed only when the view is
-    called.
+   The index **maintains and incrementally updates** the results of
+   `winning_bids` in memory. Because incremental work is performed on writes,
+   reads from indexes are computationally <redb>free</redb>.
+
+   This index can **also**  help [optimize
+   operations](/transform-data/optimization/) like point lookups and delta joins
+   on the index column(s) as well as support ad-hoc queries.
```
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md:303]: Yup.  One of those repetitions put in place to increase likelihood that people will see it.  In a future version of the docs, it'll likely be that some of the repetition can go ... but, which will be determined by our metrics and how much other pages in the docs have been updated.

File: doc/user/content/get-started/quickstart.md [hunk 8]
```diff
@@ -12,297 +12,764 @@ aliases:
...
+     SELECT *
+     FROM flippers
+   ) WITH (snapshot = false)
+   ;
    ```
 
-   You can keep an eye on the results, but these may not change much at the
-   moment. You'll fix that in the next step!
+   Because of the randomness of the data being generated, it may take some time
+   (~up to <red>1 minute</red>) for a new data to come in for users matching the
```
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md:472]: 👍 Sounds good.  Because I didn't know about the special tech that allows inputing data w/o opening a new browser window (thanks for the info!), I had used this example first because I didn't want people to open a new window.   But knowing that, I can replace this example with the one below.  This works out well because I did also want to have an optional exercise of rewriting one of the queries as a cte.

File: doc/user/content/get-started/quickstart.md [hunk 9]
```diff
@@ -12,297 +12,764 @@ aliases:
...
     SELECT * FROM winning_bids
     WHERE item = 'Best Pizza in Town'
-    ORDER BY bid_time DESC;
+    ORDER BY bid_time DESC
+    LIMIT 10;
     ```
 
-   Yikes! In Materialize, you use [**indexes**](/sql/create-index/) to
-   keep results incrementally updated and immediately accessible.
+   The queries should be faster as the reads are served from memory.
```
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md]: Updated in patch.

File: doc/user/content/get-started/quickstart.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md]: Updated in patch.

File: doc/user/content/get-started/quickstart.md (hunk 1 above)
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md]: Updated in patch.

File: doc/user/content/get-started/quickstart.md [hunk 10]
```diff
@@ -12,297 +12,764 @@ aliases:
...
+
+   - Maintain and incrementally update view results in memory; and
+
+   - Further [optimize query performance](/transform-data/optimization/) by
+     supporting point lookups on the indexed field.
+
+## Step 4. Create an index on the view
+
+In Materialize, you can create [indexes](/concepts/indexes/) on views. In
+Materialize, indexes maintain and, as data changes (inserts/updates/deletes),
```
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md]: Updated in patch.

File: doc/user/content/get-started/quickstart.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md]: Updated in patch.

File: doc/user/content/get-started/quickstart.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md]: Updated in patch.

File: doc/user/content/get-started/quickstart.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md:262]: Updated in patch.

File: doc/user/content/get-started/quickstart.md [hunk 11]
```diff
@@ -12,299 +12,703 @@ aliases:
...
+   specifies that the generator should emit new data every 1s.
 
     ```mzsql
     CREATE SOURCE auction_house
     FROM LOAD GENERATOR AUCTION
     (TICK INTERVAL '1s', AS OF 100000)
     FOR ALL TABLES;
     ```
-1. Use the [`SHOW SOURCES`](/sql/show-sources/) command to get an idea of the data being generated:
+
```
Comment (kay-kim) [inline doc/user/content/get-started/quickstart.md:103]: Wondering if I should just get rid of `, AS OF 100000`

File: doc/user/content/get-started/quickstart.md [hunk 12]
```diff
@@ -12,299 +12,703 @@ aliases:
...
+      ```
+
+   1. Run a query on `flippers`.
+
+      ```mzsql
+      SELECT *
+      FROM flippers
+      WHERE flipper_id = 450;
+      ```
+
```
Comment (kay-kim) [inline doc/user/content/get-started/quickstart.md:619]: Thinking of adding that it might be that index on the `winning_bids` view might still be better, but that it all depends.

File: doc/user/content/get-started/quickstart.md [hunk 13]
```diff
@@ -12,299 +12,703 @@ aliases:
...
+  such, an existing index may be sufficient to support your queries.
 
-As the auction house operator, you should now have a high degree of confidence that Materialize can help you implement and automate operational use cases that depend on fresh, consistent results served in an interactive way.
+- If you have created various views to act as the building blocks for a view
+  from which you will serve the results, creating an index only on the serving
+  view may be sufficient depending upon your query patterns and usage.
 
-Once you’re done exploring the auction house source, remember to clean up your environment:
+- Even without indexes, Materialize can optimize computations for queries.
+  However, this underlying work is discarded after each query run.
```
Comment (kay-kim) [inline doc/user/content/get-started/quickstart.md:691]: Just leaving a note (github PRs were having an issue last night) but there will be a separate PR to mat view vs view+index ... and once that's done, in this considerations section, will add a little blurb to point users to that.

File: doc/user/content/get-started/quickstart.md [hunk 14]
```diff
@@ -43,323 +59,740 @@ Materialize.
...
+   ORDER BY a.id,
+     b.amount DESC,
+     b.bid_time,
+     b.buyer;
+   ```
+
+   [`SELECT DISTINCT ON
+   (...)`](https://www.postgresql.org/docs/current/sql-select.html#SQL-DISTINCT)
+   is a PostgreSQL expression that keeps only the first row of each set of
+   matching rows.
```
Comment (kay-kim) [inline doc/user/content/get-started/quickstart.md:253]: Will link to idiomatic materialize sql once https://github.com/MaterializeInc/materialize/pull/29643 is merged.

File: doc/user/content/get-started/quickstart.md [hunk 15]
```diff
@@ -43,323 +59,740 @@ Materialize.
...
+   ```
+
+1. Select your schema.
+
+   ```mzsql
+   -- Replace <schema> with the name for your schema
+   SET SCHEMA <schema>;
+   ```
+
+## Step 2. Create sources
```
[the remark this answers is not in the document]
Comment (JLDLaughlin) [reply doc/user/content/get-started/quickstart.md:94]: nit part 2: some steps have punctuation while others don't, should we standardize on one or another?

File: doc/user/content/get-started/quickstart.md [hunk 16]
```diff
@@ -43,323 +59,740 @@ Materialize.
...
+      | ----- | ------ | ------------------ | -------------------------- | ------ | ----- | ---------- | ------ | -------------------------- |
+      | 15575 | 158    | Signed Memorabilia | 2024-07-25 20:30:25.085+00 | 155751 | 215   | 15575      | 27     | 2024-07-25 20:30:16.085+00 |
+      | 15575 | 158    | Signed Memorabilia | 2024-07-25 20:30:25.085+00 | 155750 | 871   | 15575      | 63     | 2024-07-25 20:30:15.085+00 |
+      | 15575 | 158    | Signed Memorabilia | 2024-07-25 20:30:25.085+00 | 155752 | 2608  | 15575      | 16     | 2024-07-25 20:30:17.085+00 |
+      ```
+
+      As new data enters the system, above `SELECT` queries must be rerun to get
+      the recent results.
+
+      By creating a view to save repeated queries and then <redb>indexing</redb>
```
[the remark this answers is not in the document]
Comment (JLDLaughlin) [reply doc/user/content/get-started/quickstart.md:212]: we had a bit of text like this in the former version:

> You can query the view directly, but this shouldn’t be very impressive just yet! Querying the view re-runs the embedded statement, which comes at some cost on growing amounts of data.

and 

> Yikes! In Materialize, you use [indexes](https://materialize.com/docs/sql/create-index/) to keep results incrementally updated and immediately accessible.

File: doc/user/content/get-started/quickstart.md [hunk 17]
```diff
@@ -12,19 +12,35 @@ aliases:
...
+- Create an [index](/concepts/indexes "Indexes represents query results stored
+  in memory within a cluster") to compute and store view results in memory. As
+  new auction data arrives, the index <redb>incrementally updates</redb> view
+  results instead of recalculating the results from scratch, making fresh
+  results immediately available for reads.
+
+- Create and query views to verify that Materialize always serves
+  <redb>consistent results</redb>.
+
+The quickstart uses the default `quickstart` cluster. In Materialize, a
```
Comment (JLDLaughlin) [inline doc/user/content/get-started/quickstart.md:40]: This line also feels a little out of place for me! Maybe it was meant to be alongside the other one?

File: doc/user/content/get-started/quickstart.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (JLDLaughlin) [reply doc/user/content/get-started/quickstart.md:25]: I think it'd be nice to tie in the motivation of the user a bit more here, particularly if we're moving away from calling out the product principles. So something roughly like:

> In this quickstart, we'll a sample auction house data set to build an operational use case around auctions, bidders and (gasp) fraud. To do that, you will:
> - Create an index to compute, and incrementally update, your SQL results. We will use this to catch the fraudster in action!

File: doc/user/content/get-started/quickstart.md [hunk 18]
```diff
@@ -12,19 +12,35 @@ aliases:
...
-This quickstart will get you up and running in a few minutes and with no
-dependencies, so you can experience the superpowers of an operational data
-store first-hand:
+Materialize provides always-fresh results while also providing [strong
+consistency guarantees](/get-started/isolation-level/). In Materialize, both
+[indexes](/concepts/indexes/ "Indexes represents query results stored in memory
+within a cluster") and [materialized views](/concepts/views/#materialized-views)
+**incrementally update** results when Materialize ingests new data; i.e., work
+is performed on writes. Because work is performed on writes, reads from these
+objects return up-to-date results while being computationally <redb>free</redb>.
```
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md:23]: 👍 Will bold -- since we tend to use purple for links.

File: doc/user/content/get-started/quickstart.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md:94]: - Our docs seem to use sentence casing ... so, will leave sources lower-case for now (but ... the casing for sections is something that we can revisit).

- Will update to use punctuations to be consistent on this page.

File: doc/user/content/get-started/quickstart.md [hunk 19]
```diff
@@ -12,299 +12,703 @@ aliases:
...
 
     ```mzsql
     CREATE SOURCE auction_house
     FROM LOAD GENERATOR AUCTION
     (TICK INTERVAL '1s', AS OF 100000)
     FOR ALL TABLES;
     ```
-1. Use the [`SHOW SOURCES`](/sql/show-sources/) command to get an idea of the data being generated:
+
+    `CREATE SOURCE` can create **multiple** tables (referred to as `subsources`
```
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md:117]: Yup.  A little repetitive .. but as people are going through the steps, I think they'd probably expect to see info about the command they ran within the step.  

- This description is more about `create source` - with an aside for subsources.
- The next description is more about what the users will see when they run the `show sources`. 

Screenshot 2024-09-23 at 11 11 18 AM

File: doc/user/content/get-started/quickstart.md [hunk 20]
```diff
@@ -43,323 +59,740 @@ Materialize.
...
+    * To view the relationship between `auctions` and `bids`, you can join by
+      the auction id. The following query returns the most recently ended
+      auctions and associated bids (limited to 3 rows):
+
+      ```mzsql
+      SELECT a.*, b.*
+      FROM auctions AS a
+      JOIN bids AS b
+        ON a.id = b.auction_id
+      ORDER BY a.end_time desc
```
Comment (kay-kim) [inline doc/user/content/get-started/quickstart.md:194]: This one ... I think I want it ordered.

Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md:194]: Actually, will remove the ordering here as well since we're only getting a sample data in these tables.

File: doc/user/content/get-started/quickstart.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md:212]: We'll add your suggestions to the next section on views.  And get rid of most here.

File: doc/user/content/get-started/quickstart.md [hunk 21]
```diff
@@ -43,323 +59,740 @@ Materialize.
...
+
+       The queries should be faster.
+
+       The above exercise is to illustrate how to create an index on
+       sources. In practice, you may find that you rarely need to index a
+       source.
+
+## Step 8. Clean up
+
+To clean up the quickstart environment:
```
[the remark this answers is not in the document]
Comment (kay-kim) [reply doc/user/content/get-started/quickstart.md:584]: Will defer on this for now.  We have it in the  lines 629 and 638.  

Reason: "You get billed for clusters..." -  people may question whether they are going to be billed even though they are using free trial.

But, we can definitely revisit.
